src/Pact/Parse.hs

```diff
@@ -31,64 +31,43 @@ import Control.Applicative
 import Data.List
 import Control.Monad
 import Prelude
-import Data.String
 import Data.Decimal
 import qualified Data.Attoparsec.Text as AP
 import Data.Char (digitToInt)
+import Data.Text (Text)
 
-import Pact.Types.Lang
+import Pact.Types.Exp
 import Pact.Types.Parser
+import Pact.Types.Info
 
 type P a = forall m. (Monad m,TokenParsing m,CharParsing m,DeltaParsing m) => PactParser m a
 
 
 -- | Main parser for Pact expressions.
-expr :: P Exp
+expr :: P (Exp Parsed)
 expr = do
   delt <- position
   let inf = do
         end <- position
         let len = bytes end - bytes delt
         return $ Parsed delt (fromIntegral len)
+      separator t s = symbol t >> ((ESeparator . SeparatorExp s) <$> inf)
   msum
-    [ TF.try (ELiteral <$> token number <*> inf) <?> "number"
-    , ELiteral . LString <$> stringLiteral <*> inf <?> "string"
-    , ELiteral <$> bool <*> inf <?> "bool"
-    , ESymbol <$> (symbolic '\'' >> ident style) <*> inf <?> "symbol"
-    , do
-      a <- ident style
-      msum
-        [ let p = do
-                t <- typed
-                EAtom a Nothing (Just t) <$> inf
-          in TF.try p <?> "typed atom"
-        , let p = do
-                q <- qualified
-                EAtom a (Just q) Nothing <$> inf
-          in TF.try p <?> "qual atom"
-        , EAtom a Nothing Nothing <$> inf <?> "bare atom"
-        ]
-    , EList <$> parens (many expr) <*> pure Nothing <*> inf <?> "sexp"
-    , do
-      is <- msum
-        [ TF.try $ brackets (many expr) <?> "space-delimited list literal"
-        , brackets (expr `sepBy` comma) <?> "comma-delimited list literal"
-        ]
-      let lty = case nub (map expPrimTy is) of
-                  [Just ty] -> ty
-                  _ -> TyAny
-      i <- inf
-      return $ EList is (Just lty) i
-    , do
-      ps <- pairs
-      let (ops, kvs) = unzip ps
-      if
-        | all (== ":") ops  -> EObject kvs <$> inf
-        | all (== ":=") ops -> EBinding kvs <$> inf
-        | otherwise         -> unexpected $ "Mixed binding/object operators: " ++ show ops
+    [ TF.try (ELiteral <$> (LiteralExp <$> token number <*> inf)) <?> "number"
```

Comment:
Also from the sexp parsing work I did, I'd like to promote the idea that rather than using `msum`, we can dispatch directly based on the first character: 

https://github.com/kadena-io/pact/blob/399d9bd8048ed60a5aa3d0e73e2eae6c31ad7a0e/src/Pact/Types/SExp.hs#L81-L99

Reply:
My goal at least in ExpParser was the opposite actually, to be more combinator-oriented not less -- combinators with `<|>` are idiomatic and flexible. What is the advantage of this dispatch style?

Reply:
My only reason for going that route was performance. `asum` / `msum` need to try every parser in order, whereas this does a one-character lookahead and jumps to the right parser (note that neither approach has to do backtracking except for in the `number` case, so this is a micro-optimization). Sure this is a bit lower-level, but for a language this small I think it's very readable.

Reply:
Did you measure a performance improvement? I ask because IMO parsers, especially atto and MP, are designed to handle failures quickly -- indeed I suspect it's the whole point of the otherwise horrifically unintuitive 4-continuation datastructure they all share. The "trivial failure" continuation which would be called here effortlessly rolls back stream state (while still accumulating `expected` errors, which would be pretty ungainly to achieve here).

Reply:
This might well be faster, switching to `case`-based dispatch served to speed up Compile parsing. We need to do a performance pass on parse and compile at some point where I would recommend trying this approach and measuring any improvement